### PR TITLE
Optimizing overload resolution [ci: last-only]

### PR DESCRIPTION
**Pull request by retronym**
Trying to reclaim some performance conceded in #6871

### CODE REVIEW

File: src/reflect/scala/reflect/internal/Types.scala [hunk 1]
```diff
@@ -4407,7 +4407,19 @@ trait Types
...
+    while (!(i.isEmpty || j.isEmpty)) {
+      if (!(i.head =:= j.head))
+        return false
+      i = i.tail
+      j = j.tail
+    }
+    i.isEmpty && j.isEmpty
+  }
 
   private var _basetypeRecursions: Int = 0
```
Comment (viktorklang) [inline src/reflect/scala/reflect/internal/Types.scala:4424]: private[this]?

Comment (retronym) [reply src/reflect/scala/reflect/internal/Types.scala:4424]: #7209

File: src/reflect/scala/reflect/internal/Definitions.scala [hunk 2]
```diff
@@ -942,40 +943,45 @@ trait Definitions extends api.StandardDefinitions {
       // look at erased type because we (only) care about what ends up in bytecode
       // (e.g., an alias type is fine as long as is compiles to a single-abstract-method)
       val tpSym: Symbol = erasure.javaErasure(tp).typeSymbol
-
-      if (tpSym.exists && tpSym.isClass && !(tpSym hasFlag (FINAL | SEALED))
+      val info = tp.typeSymbol.info
```
Comment (adriaanm) [inline src/reflect/scala/reflect/internal/Definitions.scala]: I don't see why we couldn't do this -- the symbol and its info should tell us all we need to know. Running test suite locally with `tpSym.info` as RHS.

**Comment by adriaanm** [comment]
Rebased and added a commit. I experimented a bit, but shied away from the rabbit hole. I think we can go ahead with the caching. I do worry a bit about the failing assert and discrepancies between samOf during typer and erasure.

File: src/reflect/scala/reflect/internal/Definitions.scala [hunk 3]
```diff
@@ -931,58 +931,63 @@ trait Definitions extends api.StandardDefinitions {
...
-          // (implementation restriction: implicit argument lists are excluded to simplify type inference in adaptToSAM)
-          && {
-          val ctor = tpSym.primaryConstructor
-          !ctor.exists || (!ctor.isOverloaded && ctor.isPublic && ctor.info.params.isEmpty && ctor.info.paramSectionCount <= 1)
+      *
+      * The method must be monomorphic and have exactly one parameter list.
+      * The class defining the method is a supertype of `tp` that
+      * has a public no-arg primary constructor and it can be subclassed (not final or sealed).
+      *
+      * Note that this is also used during erasure (TODO: maybe we could simplify typedFunction for post-typer usage and avoid this?),
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Definitions.scala:941]: >  maybe we could simplify typedFunction for post-typer usage and avoid this?

👍 , we shouldn't be doing SAM lookups post typer IMO, but instead should use the `SAMFunction` attachment of a previously typed `Function` to remember what we computed originally.

We already do some of that in `ErasureTyper.typed1`:

https://github.com/scala/scala/blob/ab66e2618c4bc6c3fb6c335ff4f63edb6514b3df/src/compiler/scala/tools/nsc/transform/Erasure.scala#L813-L817

Comment (retronym) [reply src/reflect/scala/reflect/internal/Definitions.scala:941]: The `deferredMembers.head.info.paramSectionCount == 1` condition can change a class from being a SAM before/after uncurry, for example:

```
scala> trait Test { def apply(a: Int)(b: Int): Int }
defined trait Test

scala> exitingTyper(samOf(typeOf[Test]))
res4: $r.intp.global.Symbol = <none>

scala> exitingUncurry(samOf(typeOf[Test]))
res5: $r.intp.global.Symbol = method apply
```

Comment (adriaanm) [reply src/reflect/scala/reflect/internal/Definitions.scala:941]: I also saw multiple deferred members show up, that should really be overrides (covariant refined return types)

**Pull request merged**